From f03bb30c84c9ed6962878562d3655cdae4a96bb7 Mon Sep 17 00:00:00 2001
From: l4h <l4h@localhost>
Date: Wed, 28 Oct 2026 10:15:58 -0700
Subject: [PATCH] x86_64: preserve syscall RSP with the SKIM window

Save user RSP before switching to the kernel page tables, and carry the
return RSP through existing per-CPU SKIM scratch storage. Clear the scratch
word after every read, including the interrupt-return RCX path. Handle UP
as well as SMP, and the normal and IPC fastpath returns.

Allow trapSyscalls with SKIM now that Linux syscall entry and return
preserve the stack pointer without exposing the TCB or dereferencing the
user stack.
---
 .../arch/x86/arch/64/mode/fastpath/fastpath.h | 23 ++++++++++++++--
 include/arch/x86/arch/64/mode/model/smp.h     |  8 +++---
 .../arch/x86/arch/64/mode/model/statedata.h   |  3 +++
 libsel4/include/sel4/constants.h              |  4 ---
 src/arch/x86/64/c_traps.c                     | 26 +++++++++++++++++--
 src/arch/x86/64/model/statedata.c             |  6 ++---
 src/arch/x86/64/traps.S                       | 23 +++++++++++++---
 7 files changed, 74 insertions(+), 19 deletions(-)

diff --git a/include/arch/x86/arch/64/mode/fastpath/fastpath.h b/include/arch/x86/arch/64/mode/fastpath/fastpath.h
index ae851a334..8ae202bd3 100644
--- a/include/arch/x86/arch/64/mode/fastpath/fastpath.h
+++ b/include/arch/x86/arch/64/mode/fastpath/fastpath.h
@@ -150,6 +150,14 @@ static inline void NORETURN FORCE_INLINE fastpath_restore(word_t badge, word_t m
     restore_user_debug_context(cur_thread);
 #endif
 
+#ifdef CONFIG_KERNEL_SKIM_WINDOW
+#ifdef ENABLE_SMP_SUPPORT
+    word_t *skim_scratch = (word_t *)nodeSkimScratch[CURRENT_CPU_INDEX()];
+#else
+    word_t *skim_scratch = &x64KSSyscallUserRSP;
+#endif
+#endif
+
 #ifdef ENABLE_SMP_SUPPORT
 #ifdef CONFIG_KERNEL_SKIM_WINDOW
     word_t next_cr3 = MODE_NODE_STATE(x64KSCurrentUserCR3);
@@ -217,6 +225,9 @@ static inline void NORETURN FORCE_INLINE fastpath_restore(word_t badge, word_t m
             : "memory"
         );
     } else {
+#ifdef CONFIG_KERNEL_SKIM_WINDOW
+        *skim_scratch = getRegister(cur_thread, RSP);
+#endif
         asm volatile(
             // Set our stack pointer to the top of the tcb so we can efficiently pop
             "movq %0, %%rsp\n"
@@ -246,8 +257,15 @@ static inline void NORETURN FORCE_INLINE fastpath_restore(word_t badge, word_t m
 #endif /* CONFIG_KERNEL_SKIM_WINDOW */
 #endif /* defined(ENABLE_SMP_SUPPORT) && defined(CONFIG_KERNEL_SKIM_WINDOW) */
 #ifdef CONFIG_KERNEL_SKIM_WINDOW
-            // clear RSP to not leak information to the user
-            "xor %%rsp, %%rsp\n"
+#ifdef ENABLE_SMP_SUPPORT
+            "swapgs\n"
+            "movq %%gs:%c[syscall_scratch], %%rsp\n"
+            "movq $0, %%gs:%c[syscall_scratch]\n"
+            "swapgs\n"
+#else
+            "movq (x64KSSyscallUserRSP), %%rsp\n"
+            "movq $0, (x64KSSyscallUserRSP)\n"
+#endif
 #else
             // restore the thread's RSP (saved on syscall entry)
             "movq 8(%%rsp), %%rsp\n"
@@ -261,6 +279,7 @@ static inline void NORETURN FORCE_INLINE fastpath_restore(word_t badge, word_t m
             "S"(msgInfo)
 #if defined(ENABLE_SMP_SUPPORT) && defined(CONFIG_KERNEL_SKIM_WINDOW)
             , "c"(next_cr3)
+            , [syscall_scratch] "i"(nodeSkimScratchOffset)
 #endif
             : "memory"
         );
diff --git a/include/arch/x86/arch/64/mode/model/smp.h b/include/arch/x86/arch/64/mode/model/smp.h
index 61a420d87..3fce8fda0 100644
--- a/include/arch/x86/arch/64/mode/model/smp.h
+++ b/include/arch/x86/arch/64/mode/model/smp.h
@@ -37,10 +37,10 @@ compile_assert(nodeInfoIsCacheSized, (sizeof(nodeInfo_t) % L1_CACHE_LINE_SIZE) =
 extern nodeInfo_t node_info[CONFIG_MAX_NUM_NODES] ALIGN(L1_CACHE_LINE_SIZE);
 
 #ifdef CONFIG_KERNEL_SKIM_WINDOW
-/* we only need 1 word of scratch space, which we know we will at least get by the size
- 8 of the nodeInfo_t struct so we define this array as char to ensure it is sized correctly.
- We need each element of the array to be precisely the same size as each element of node_info
- so that the offset between the element of each array is a constant */
+/* One word per CPU, used for RSP on syscall entry/return and RCX on
+ * interrupt return. Clear it immediately after reading it: this window
+ * remains mapped in every user address space. Matching node_info's stride
+ * lets assembly address it through GS without mapping node_info itself. */
 extern char nodeSkimScratch[CONFIG_MAX_NUM_NODES][sizeof(nodeInfo_t)] ALIGN(L1_CACHE_LINE_SIZE) VISIBLE SKIM_BSS;
 compile_assert(nodeInfoAndScratchSameSize, sizeof(node_info) == sizeof(nodeSkimScratch))
 /* this will be declared in the linker script as the offset between node_info and
diff --git a/include/arch/x86/arch/64/mode/model/statedata.h b/include/arch/x86/arch/64/mode/model/statedata.h
index e5f4d8c72..9fd1b8381 100644
--- a/include/arch/x86/arch/64/mode/model/statedata.h
+++ b/include/arch/x86/arch/64/mode/model/statedata.h
@@ -49,3 +49,6 @@ NODE_STATE_END(modeNodeState);
 #define IRQ_STACK_SIZE 6
 extern word_t x64KSIRQStack[CONFIG_MAX_NUM_NODES][IRQ_STACK_SIZE + 2] ALIGN(64) VISIBLE SKIM_BSS;
 
+#ifndef ENABLE_SMP_SUPPORT
+extern word_t x64KSSyscallUserRSP VISIBLE SKIM_BSS;
+#endif
diff --git a/libsel4/include/sel4/constants.h b/libsel4/include/sel4/constants.h
index 831db07f9..1b8741089 100644
--- a/libsel4/include/sel4/constants.h
+++ b/libsel4/include/sel4/constants.h
@@ -95,11 +95,7 @@ typedef enum {
 
     SEL4_FORCE_LONG_ENUM(seL4_TCBFlag),
     seL4_TCBFlag_MASK = seL4_TCBFlag_NoFlag
-#if !(defined(CONFIG_ARCH_X86_64) && defined(CONFIG_KERNEL_SKIM_WINDOW))
-                        /* The SKIM window's sysret clears the user RSP,
-                         * which trapping threads must get back. */
                         | seL4_TCBFlag_trapSyscalls
-#endif
 #ifdef CONFIG_HAVE_FPU
                         | seL4_TCBFlag_fpuDisabled
 #endif
diff --git a/src/arch/x86/64/c_traps.c b/src/arch/x86/64/c_traps.c
index a1c4e5ea8..cdb4f08e5 100644
--- a/src/arch/x86/64/c_traps.c
+++ b/src/arch/x86/64/c_traps.c
@@ -238,6 +238,14 @@ void VISIBLE NORETURN restore_user_context(void)
     restore_user_debug_context(cur_thread);
 #endif
 
+#ifdef CONFIG_KERNEL_SKIM_WINDOW
+#ifdef ENABLE_SMP_SUPPORT
+    word_t *skim_scratch = (word_t *)nodeSkimScratch[CURRENT_CPU_INDEX()];
+#else
+    word_t *skim_scratch = &x64KSSyscallUserRSP;
+#endif
+#endif
+
 #ifdef ENABLE_SMP_SUPPORT
 #ifdef CONFIG_KERNEL_SKIM_WINDOW
     word_t user_cr3 = MODE_NODE_STATE(x64KSCurrentUserCR3);
@@ -331,6 +339,11 @@ void VISIBLE NORETURN restore_user_context(void)
                 : "memory"
             );
         } else {
+#ifdef CONFIG_KERNEL_SKIM_WINDOW
+            /* The TCB disappears when CR3 changes. Only the returning
+             * thread's RSP crosses that switch through the SKIM window. */
+            *skim_scratch = getRegister(cur_thread, RSP);
+#endif
             asm volatile(
                 // Set our stack pointer to the top of the tcb so we can efficiently pop
                 "movq %0, %%rsp\n"
@@ -362,8 +375,15 @@ void VISIBLE NORETURN restore_user_context(void)
 #endif /* CONFIG_KERNEL_SKIM_WINDOW */
 #endif /* defined(ENABLE_SMP_SUPPORT) && defined(CONFIG_KERNEL_SKIM_WINDOW) */
 #ifdef CONFIG_KERNEL_SKIM_WINDOW
-                // clear RSP to not leak information to the user
-                "xor %%rsp, %%rsp\n"
+#ifdef ENABLE_SMP_SUPPORT
+                "swapgs\n"
+                "movq %%gs:%c[syscall_scratch], %%rsp\n"
+                "movq $0, %%gs:%c[syscall_scratch]\n"
+                "swapgs\n"
+#else
+                "movq (x64KSSyscallUserRSP), %%rsp\n"
+                "movq $0, (x64KSSyscallUserRSP)\n"
+#endif
 #else
                 // restore the thread's RSP (saved on syscall entry)
                 "movq 8(%%rsp), %%rsp\n"
@@ -375,6 +395,7 @@ void VISIBLE NORETURN restore_user_context(void)
                 : "r"(&cur_thread->tcbArch.tcbContext.registers[RDI])
 #if defined(ENABLE_SMP_SUPPORT) && defined(CONFIG_KERNEL_SKIM_WINDOW)
                 , "c"(user_cr3)
+                , [syscall_scratch] "i"(nodeSkimScratchOffset)
 #endif
                 // Clobber memory so the compiler is forced to complete all stores
                 // before running this assembler
@@ -435,6 +456,7 @@ void VISIBLE NORETURN restore_user_context(void)
              * we stashed */
             "movq %%rcx, %%cr3\n"
             "movq %%gs:%c[scratch_offset], %%rcx\n"
+            "movq $0, %%gs:%c[scratch_offset]\n"
 #endif /* CONFIG_KERNEL_SKIM_WINDOW */
             "addq $8, %%rsp\n"
             // Switch to the user GS value
diff --git a/src/arch/x86/64/model/statedata.c b/src/arch/x86/64/model/statedata.c
index 1fb6935a7..b012565fb 100644
--- a/src/arch/x86/64/model/statedata.c
+++ b/src/arch/x86/64/model/statedata.c
@@ -30,7 +30,7 @@ UP_STATE_DEFINE(cr3_t, x64KSCurrentCR3);
 
 word_t x64KSIRQStack[CONFIG_MAX_NUM_NODES][IRQ_STACK_SIZE + 2] ALIGN(64) VISIBLE SKIM_BSS;
 
-#if !defined(ENABLE_SMP_SUPPORT) && !defined(CONFIG_KERNEL_SKIM_WINDOW)
-/* Scratch slot for the user RSP on 'syscall' entry (see traps.S). */
-word_t x64KSSyscallUserRSP VISIBLE;
+#ifndef ENABLE_SMP_SUPPORT
+/* Accessible across the SKIM address-space switch; cleared after use. */
+word_t x64KSSyscallUserRSP VISIBLE SKIM_BSS;
 #endif
diff --git a/src/arch/x86/64/traps.S b/src/arch/x86/64/traps.S
index 3ead5c93a..4c79efdc3 100644
--- a/src/arch/x86/64/traps.S
+++ b/src/arch/x86/64/traps.S
@@ -600,9 +600,26 @@ END_FUNC(kernel_exception)
 # the current RFLAGS (after saving) have been masked
 # with IA32_FMASK.
 BEGIN_FUNC(handle_fastsyscall)
-    LOAD_KERNEL_AS(rsp)
     MAYBE_SWAPGS
-#ifndef CONFIG_KERNEL_SKIM_WINDOW
+#ifdef CONFIG_KERNEL_SKIM_WINDOW
+    /* SYSCALL leaves user RSP intact. Save it before LOAD_KERNEL_AS uses
+     * RSP as a temporary; node_info and the TCB are not mapped yet. */
+#ifdef ENABLE_SMP_SUPPORT
+    movq    %rsp, %gs:nodeSkimScratchOffset
+#else
+    movq    %rsp, x64KSSyscallUserRSP
+#endif
+    LOAD_KERNEL_AS(rsp)
+    LOAD_USER_CONTEXT
+    addq    $8, %rsp
+#ifdef ENABLE_SMP_SUPPORT
+    pushq   %gs:nodeSkimScratchOffset
+    movq    $0, %gs:nodeSkimScratchOffset
+#else
+    pushq   x64KSSyscallUserRSP
+    movq    $0, x64KSSyscallUserRSP
+#endif
+#else
     /* Save the user RSP, which 'syscall' leaves untouched, into the RSP
      * slot of the user context so that the thread's stack pointer is
      * known to fault handlers and restored on return (threads that do
@@ -618,8 +635,6 @@ BEGIN_FUNC(handle_fastsyscall)
     addq    $8, %rsp
     pushq   x64KSSyscallUserRSP
 #endif
-#else
-    LOAD_USER_CONTEXT
 #endif
     pushq   $-1             # set Error -1 to mean entry via syscall
     push    %rcx            # save NextIP
-- 
2.47.3

